### PR TITLE
Auto sync parent terms

### LINKED ISSUE #3105: BUG: Parent taxonomy term auto sync bug
**Describe the bug**

When linking a child taxonomy term with a post, ElasticPress will add the parent and child terms info to the post document in Elasticsearch, but when edit the info of the parent taxonomy term for example changing its slug it will not be synced automatically to the post document in Elasticsearch unlike when editing the child taxonomy term slug it will synced automatically.

Because of that if I queried posts via the new parent taxonomy term slug it will not work until I make a full manual sync via Sync feature of ElasticPress because it will fix the issue via update the parent term info.

**Steps to Reproduce**

1. Go to WordPress default categories
2. Add 2 terms, one of them child to the other
3. Go to WordPress default posts
4. Add a new post and link the child term with it
5. Edit the parent term slug
6. Go to Elasticsearch and check the post document and you will find the old parent term slug is still exist
7. Optional: Make a full sync via sync feature of ElasticPress and check the post document again and everything will be ok

**Expected behavior**

The parent taxonomy term info should be synced automatically when it's got changed

**Screenshots**

**Environment information**
 - Device: Laptop 
 - OS: Ubuntu 22.04 
 - Browser and version: Chrome 107.0.5304.87 
 - WordPress version: 6.0.1 
 
 - ElasticPress version: 4.3.1 
 - Elasticsearch version: 7.10.2 
 - Where do you host your Elasticsearch server? Local docker setup 
 - What ElasticPress features do you have currently active? Post Search, Custom Search Results, Related Posts, Protected Content, Terms 
 - Plugins and version: 
 - Theme and version: 
 
 - <details><summary>Site Health Info:</summary>
   
   </details>

**Additional context**

### PR DESCRIPTION
**Pull request by MARQAS**
### Description of the Change
Auto-sync the parent terms when child terms are selected in the posts

Closes #3105 

### How to test the Change

- Go to WordPress default categories
- Add 2 terms, make one of them child to the other
- Go to WordPress default posts
- Add a new post and link the child term with it
- Edit the parent term slug
- Go to Elasticsearch and check the post document and note that parent slug is also updated

### Changelog Entry

> Fixed - Auto sync issue with parent term
### Credits

Props @MARQAS 

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

**Comment by felipeelia** [comment]
@MARQAS a couple of questions:
1. Can we add some unit tests for this?
2. Do we also have the bug when a parent term is deleted? If so, this covers that scenario as well?
3. To increase readability, do you mind wrapping variables in double quote strings with `{}`, like in `"SELECT object_id FROM {$wpdb->term_relationships} WHERE term_taxonomy_id...`
4. Will we have a performance problem if too many posts are associated with child terms?

File: tests/php/indexables/TestPost.php [hunk 1]
```diff
@@ -7382,6 +7382,48 @@ public function testPostEditedTerm() {
...
+		$term1       = wp_insert_term( $term_1_name, $tax_name );
+
+		$term_2_name = rand_str( 32 );
+		$term2       = wp_insert_term( $term_2_name, $tax_name, array( 'parent' => $term1['term_id'] ) );
+
+		wp_set_object_terms( $post->ID, array( $term2['term_id'] ), $tax_name, true );
+
+		ElasticPress\Elasticsearch::factory()->refresh_indices();
+
+		$test_tag = get_term_by( 'id', $term1['term_id'], $tax_name );
```
Comment (felipeelia) [inline tests/php/indexables/TestPost.php]:
```suggestion
		$term_1_name = rand_str( 32 );
		$term_1      = wp_insert_term( $term_1_name, $tax_name );

		$term_2_name = rand_str( 32 );
		$term_2      = wp_insert_term( $term_2_name, $tax_name, array( 'parent' => $term_1['term_id'] ) );

		wp_set_object_terms( $post->ID, array( $term_2['term_id'] ), $tax_name, true );

		ElasticPress\Elasticsearch::factory()->refresh_indices();

		$test_tag = get_term_by( 'id', $term_1['term_id'], $tax_name );
```

File: tests/php/indexables/TestPost.php [hunk 2]
```diff
@@ -7382,6 +7382,48 @@ public function testPostEditedTerm() {
...
+				'name' => 'Parent Term',
+			]
+		);
+
+		ElasticPress\Indexables::factory()->get( 'post' )->sync_manager->index_sync_queue();
+		ElasticPress\Elasticsearch::factory()->refresh_indices();
+
+		$document = ElasticPress\Indexables::factory()->get( 'post' )->get( $post->ID );
+		$this->assertEquals( 'parent-term', $document['terms'][$tax_name][1]['slug'] );
+		$this->assertEquals( 'Parent Term', $document['terms'][$tax_name][1]['name'] );
```
Comment (felipeelia) [inline tests/php/indexables/TestPost.php]:
```suggestion
		$this->assertEquals( 'parent-term', $document['terms'][ $tax_name ][1]['slug'] );
		$this->assertEquals( 'Parent Term', $document['terms'][ $tax_name ][1]['name'] );
```

**Review by felipeelia — CHANGES_REQUESTED** [review]
@MARQAS do you mind making the suggestions I've added? Just so we keep var names consistent. Thanks!

**Comment by MARQAS** [comment]
@felipeelia I have added the changes you asked for along with a notice. I observed that it wasn't showing notice when the parent term was not attached to the posts. 
Steps to reproduce: 

- Add parent and child terms
- Attach child term to the posts
- Make `Content Items per Index Cycle` less than the posts attached to the child term.
- While editing a child term, you can see the notice. 
- But it will not show the notice when editing a parent term. 
- It will only show the notice on the parent term if the posts attached to the parent term are greater than `Content Items per Index Cycle`

With this PR, it will show the notice on the parent term as well if posts associated with one of its child terms are greater than the `Content Items per Index Cycle`

**Pull request merged**